src/compare.rs

```diff
@@ -0,0 +1,506 @@
+//! Comparators.
+//!
+//! A comparator is any type that implements the [`Compare`](trait.Compare.html) trait,
+//! which imposes a [total order](https://en.wikipedia.org/wiki/Total_order). Its
+//! [`compare`](trait.Compare.html#tymethod.compare) method accepts two values, which may
+//! be of the same type or different types, and returns an ordering on them.
+//!
+//! Comparators are useful for parameterizing the behavior of sort methods and certain
+//! data structures.
+//!
+//! The most basic comparator is [`Natural`](struct.Natural.html), which simply delegates
+//! to the type's implementation of [`Ord`]
+//! (http://doc.rust-lang.org/std/cmp/trait.Ord.html):
+//!
+//! ```rust
+//! use collect::compare::{Compare, Natural};
+//!
+//! let a = &1u8;
+//! let b = &2u8;
+//!
+//! let cmp = Natural;
+//! assert_eq!(cmp.compare(a, b), Less);
+//! assert_eq!(cmp.compare(b, a), Greater);
+//! assert_eq!(cmp.compare(a, a), Equal);
+//! ```
+//!
+//! There are convenience methods for checking each of the six relations:
+//!
+//! ```rust
+//! use collect::compare::{Compare, Natural};
+//!
+//! let a = &1u8;
+//! let b = &2u8;
+//!
+//! let cmp = Natural;
+//! assert!(cmp.compares_lt(a, b));
+//! assert!(cmp.compares_le(a, b));
+//! assert!(cmp.compares_ge(b, a));
+//! assert!(cmp.compares_gt(b, a));
+//! assert!(cmp.compares_eq(a, a));
+//! assert!(cmp.compares_ne(a, b));
+//! ```
+//!
+//! The [`CompareExt`](trait.CompareExt.html) trait provides extension methods that
+//! consume a comparator to produce a new one with different behavior, similar to
+//! [iterator adaptors](http://doc.rust-lang.org/std/iter/trait.IteratorExt.html). For
+//! example, all comparators can be [reversed](trait.CompareExt.html#method.rev):
+//!
+//! ```rust
+//! use collect::compare::{Compare, CompareExt, Natural};
+//!
+//! let cmp = Natural.rev();
```

Comment:
`Rev` does not have a default type of `Natural`, and it actually cannot have that be the case because `Natural` is itself parameterized by `Lhs` (crucial for type inference), which would have to be added as a parameter to `Rev`. You would never be able to write
```rust
let cmp = Rev;
```
anyway, because `Rev` contains a field, which is the comparator that it is reversing. We _could_ make the field of `Rev` (and the other adapters) public, like
```rust
pub struct Rev<C>(pub C);
```
and then you could write
```rust
let cmp = Rev(Natural);
```
without importing `CompareExt`, but you would still have to import `Natural` and `Rev`, obviously. One positive effect of making those fields public (and it's unlikely that the representation would ever change for adapters that are so simple) is that the adapters could be used in constant expressions. The benefit of `CompareExt`, which would presumably be in some kind of prelude once we get those, is that you don't have to `use` the individual adapter types.